### PR TITLE
Add nginx and supervisor

### PR DESCRIPTION
**Pull request by KaitCrawford**

### CODE REVIEW

**Comment by KaitCrawford** [comment]
@hodgestar please take a look

**Comment by JayH5** [comment]
Please use the praekeltfoundation/supervisor base image to do this sort of thing.

**Comment by KaitCrawford** [comment]
@JayH5 does the most recent commit look better to you?

**Comment by JayH5** [comment]
@KaitCrawford the `praekeltfoundation/supervisor` image sets up supervisor for you.

With the supervisor image, you rather just create the supervisor config files for each process you want to run... something like:  
`nginx.conf`:

``` ini
[program:nginx]
command=nginx -g "daemon off;"
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0
```

`junebug.conf`:

``` ini
[program:junebug]
command=/scripts/junebug-entrypoint.sh
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0
```

`Dockerfile`:

``` dockerfile
FROM praekeltfoundation/supervisor
MAINTAINER Praekelt Foundation <dev@praekeltfoundation.org>

RUN apt-get-install.sh gcc python-pip python-dev libjpeg-dev zlib1g-dev
RUN pip install vxyowsup
RUN pip install -q junebug
COPY ./nginx.conf /etc/supervisor/conf.d/nginx.conf
COPY ./junebug.conf /etc/supervisor/conf.d/junebug.conf
COPY ./junebug-entrypoint.sh /scripts/
EXPOSE 8080
```

Yes this is kind of a pain (especially with all those `stdout_XX` lines) but it's a lot cleaner and ensures that we have Supervisord set up the same way in all our images. FWIW, the Nginx command I got from the official Nginx Docker image [here](https://github.com/nginxinc/docker-nginx/blob/f48d943038eaafd1f69cb14d86db95b3044bd4d8/mainline/jessie/Dockerfile#L25).

**Comment by JayH5** [comment]
OK so that Dockerfile is not quite right :/ I forgot to actually install Nginx. If you ditch the yowsup stuff then you don't need to install any of that `-dev` stuff.

``` Dockerfile
FROM praekeltfoundation/supervisor
MAINTAINER Praekelt Foundation <dev@praekeltfoundation.org>

RUN apt-get-install.sh nginx
RUN pip install junebug
COPY ./nginx.conf /etc/supervisor/conf.d/nginx.conf
COPY ./junebug.conf /etc/supervisor/conf.d/junebug.conf
COPY ./junebug-entrypoint.sh /scripts/
EXPOSE 8080
```

**Comment by KaitCrawford** [comment]
I see you've removed `gcc` and `python-pip` as well (they were in the same line as the `-dev` stuff). was that accidental or intentional?

**Comment by JayH5** [comment]
@KaitCrawford: Intentional.. sorry should have been clearer. They are only needed to build native extensions for some of the yowsup stuff.

**Comment by hodgestar** [comment]
We should include the extensions needed for the yowsup stuff, I think (in order to make this a generic Junebug container that many projects can use).

**Comment by JayH5** [comment]
In that case I can build the yowsup stuff separately and add it to our pypi repo. I've done a bunch of work so that we don't end up with big Docker images with lots of build tools and compilers in them and I'd like for that work not to go wasted.

OTOH, yowsup is really a separate issue from this PR so maybe this isn't the time/place to fix that.

**Comment by hodgestar** [comment]
Building the yowsup stuff separately would be awesome.

**Comment by KaitCrawford** [comment]
@JayH5 did I get that right?
I've left the yowsup install and dependencies for now so it doesn't hold this up.

File: docker/junebug.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+command=/scripts/junebug-entrypoint.sh
+stdout_logfile=/dev/stdout
+stdout_logfile_maxbytes=0
+stderr_logfile=/dev/stderr
+stderr_logfile_maxbytes=0
```
Comment (hodgestar) [inline docker/junebug.conf]: Missing newline at EOF.

File: docker/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[nginx]
+command=nginx -g "daemon off;"
+stdout_logfile=/dev/stdout
+stdout_logfile_maxbytes=0
+stderr_logfile=/dev/stderr
+stderr_logfile_maxbytes=0
```
Comment (hodgestar) [inline docker/nginx.conf]: And here.

File: junebug-entrypoint.sh [hunk 3]
```diff
@@ -1,5 +1,42 @@
 #!/bin/bash -e
 
+echo "=> Setup Supervisord"
```
Comment (hodgestar) [inline junebug-entrypoint.sh]: I think we can revert all the changes to this file?

**Comment by KaitCrawford** [comment]
@hodgestar's comments handled

File: docker/junebug.conf [hunk 4]
```diff
@@ -0,0 +1,5 @@
+command=/scripts/junebug-entrypoint.sh
```
Comment (JayH5) [inline docker/junebug.conf:2]: Sorry, I updated my comment a few minutes after posting it... This should start with a `[program:...]` line. See the [docs](http://supervisord.org/running.html#adding-a-program).

File: docker/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[nginx]
```
Comment (JayH5) [inline docker/nginx.conf]: Same here

**Comment by JayH5** [comment]
Cool. Have you tried it out?

**Comment by hodgestar** [comment]
:+1: from me (and trying it out is a good idea).

**Comment by KaitCrawford** [comment]
@JayH5 that would be cool but I'm not sure how. I presume it would be run the image with docker but I haven't done that before and I wouldn't know what the pass conditions are. Just that it runs without any errors?

**Comment by JayH5** [comment]
@KaitCrawford: Yeah I'm not talking about extensive testing..

Just something like:

```
docker build -t junebug .
docker run --rm -it junebug
```

And see what it does... It'll probably just result in Junebug failing to start due to no redis/rabbit.

**Comment by JayH5** [comment]
FWIW, Docker is pretty cool for spinning up a bunch of services for a dev environment. If you want to try with redis/rabbit you _should_ be able to do something like:

```
docker run -d --name jb-redis redis
docker run -d --name jb-rabbitmq rabbitmq
docker run --rm -it -e REDIS_HOST=redis -e AMQP_HOST=rabbitmq --link jb-redis:redis --link jb-rabbitmq:rabbitmq junebug
```

This:
1. Starts up a redis container called `jb-redis` in the background.
2. Starts up a rabbitmq container called `jb-rabbitmq` in the background.
3. Starts up the Junebug container and links in the redis and rabbitmq containers to the Junebug container's networking.

**Comment by hodgestar** [comment]
I pushed a commit to address #3 since it's a tiny change.

**Comment by KaitCrawford** [comment]
@hodgestar, @JayH5 I added a commit to remove the installs made redundant by #4

File: Dockerfile [hunk 6]
```diff
@@ -1,16 +1,10 @@
...
 
-RUN apt-get-install.sh gcc
-RUN apt-get-install.sh python-pip
-RUN apt-get-install.sh python-dev
-RUN apt-get-install.sh libjpeg-dev
-RUN apt-get-install.sh zlib1g-dev
-RUN pip install vxyowsup
+RUN apt-get-install.sh libjpeg62
+RUN apt-get-install.sh nginx
 RUN pip install -q junebug
```
Comment (hodgestar) [inline Dockerfile:6]: We still need to install vxyowsup (I guess `RUN pip install -q vxyowsup`).

File: Dockerfile [hunk 7]
```diff
@@ -1,16 +1,10 @@
...
 MAINTAINER Praekelt Foundation <dev@praekeltfoundation.org>
 
-RUN apt-get-install.sh gcc
-RUN apt-get-install.sh python-pip
-RUN apt-get-install.sh python-dev
-RUN apt-get-install.sh libjpeg-dev
-RUN apt-get-install.sh zlib1g-dev
-RUN pip install vxyowsup
+RUN apt-get-install.sh libjpeg62
+RUN apt-get-install.sh nginx
```
Comment (JayH5) [inline Dockerfile]: Each time `apt-get-install.sh` runs, it does an `apt-get update` which can be pretty slow. Generally better to do all the installs in one line.

File: Dockerfile [hunk 8]
```diff
@@ -1,16 +1,10 @@
-FROM praekeltfoundation/python-base
+FROM praekeltfoundation/supervisor
 MAINTAINER Praekelt Foundation <dev@praekeltfoundation.org>
 
-RUN apt-get-install.sh gcc
-RUN apt-get-install.sh python-pip
-RUN apt-get-install.sh python-dev
-RUN apt-get-install.sh libjpeg-dev
-RUN apt-get-install.sh zlib1g-dev
-RUN pip install vxyowsup
```
Comment (JayH5) [inline Dockerfile:9]: You're not installing `vxyowsup` now :cry:

Comment (KaitCrawford) [inline Dockerfile:9]: whoops /sigh

**Comment by hodgestar** [comment]
:+1: from me.

**Pull request merged**